### PR TITLE
Pin the rancher-webhook chart to an exact version

### PR DESCRIPTION
**Pull request by maxsokolovsky**
## Problem
Rancher specifies the minimum version of the webhook in its configuration. When there is a newer version of the chart, Rancher will try to install it, which we don't always want, since there might be incompatibilities of Rancher and a given version of the webhook.
 
## Solution
Install only the webhook specified in configuration. Do not use the notion of a "minimum" version for the webhook chart. Instead, use exact version that doesn't change automatically for a given version of Rancher - it's to remain fixed.
 
### Manual Testing
1. Run Rancher with the webhook chart version specified with `CATTLE_RANCHER_WEBHOOK_VERSION`. Ensure that Rancher installs that version only, whether there are any newer versions or not.
1. In the UI, upgrade or downgrade to another version. Ensure that after some time (depending on the configurable chart refresh interval), Rancher switches back to the original version.
1. Rerun Rancher with a different version specified. Ensure Rancher installs that version of the webhook chart.

### Regressions Considerations
Ensure no changes occur for other system charts, like fleet. For them, the minimum version should behave as before, that is, Rancher may upgrade to a newer version if one is available.

### CODE REVIEW

File: pkg/catalogv2/system/system.go [hunk 1]
```diff
@@ -427,9 +432,10 @@ func (m *Manager) isInstalled(namespace, name, version, minVersion string, desir
 			}
 		}
 
-		if (desired.LessThan(ver) || desired.Equal(ver)) && bytes.Equal(patchedJSON, actualValueJSON) {
-			return true, "", nil, nil
+		if !desired.Equal(ver) {
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system.go:435]: 1. I think the old check should be used if minVersion is specified because, without it we are forcing other charts to always be on the latest release.
2. In this new check we need to also validate that we are installed with the expected values.

File: pkg/catalogv2/system/system.go [hunk 2]
```diff
@@ -211,13 +212,14 @@ func (m *Manager) Uninstall(namespace, name string) error {
 	return m.waitPodDone(op)
 }
 
-func (m *Manager) Ensure(namespace, name, minVersion string, values map[string]interface{}, forceAdopt bool, installImageOverride string) error {
+func (m *Manager) Ensure(namespace, name, minVersion, exactVersion string, values map[string]interface{}, forceAdopt bool, installImageOverride string) error {
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system.go:215]: This will need unit tests.

File: pkg/catalogv2/system/system.go [hunk 3]
```diff
@@ -227,22 +229,25 @@ func (m *Manager) Ensure(namespace, name, minVersion string, values map[string]i
 	return nil
 }
 
-func (m *Manager) Remove(namespace, name, minVersion string) {
-	delete(m.desiredCharts, desiredKey{
-		namespace:  namespace,
-		name:       name,
-		minVersion: minVersion,
-	})
+func (m *Manager) Remove(namespace, name string) {
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system.go:232]: This will also nee unit tests.

File: pkg/settings/setting.go [hunk 4]
```diff
@@ -167,8 +167,8 @@ var (
 	// Deprecated: On removal use kubeconfig-default-ttl-minutes for all kubeconfigs.
 	KubeconfigTokenTTLMinutes = NewSetting("kubeconfig-token-ttl-minutes", "960") // 16 hours
 
-	// RancherWebhookMinVersion is the minimum version of the webhook that rancher will install.
-	RancherWebhookMinVersion = NewSetting("rancher-webhook-min-version", "")
+	// RancherWebhookVersion is the exact version of the webhook that Rancher will install.
+	RancherWebhookVersion = NewSetting("rancher-webhook-version", "")
```
Comment (KevinJoiner) [inline pkg/settings/setting.go:171]: I don't know if we want to replace this setting in a patch release since it will break users. 
We will need to add the new setting and deprecate `RancherWebhookMinVersion`

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/catalogv2/system/system.go (hunk 3 above)
Comment (maxsokolovsky) [reply pkg/catalogv2/system/system.go:232]: This is easy to test, yes.

File: pkg/catalogv2/system/system.go (hunk 2 above)
Comment (maxsokolovsky) [reply pkg/catalogv2/system/system.go:215]: Agreed, but it's hard to test. The design of the struct will need to be changed significantly. The problem with `Ensure` is that it runs asynchronously - returns immediately. We could then test the private methods (which is already not a great solution). Specifically - `install` and `isInstalled`. They would also need to be refactored. I could create independent, small functions that only work with version strings and maybe `values` maps and report whether a version is outdated or not.

File: pkg/catalogv2/system/system.go (hunk 1 above)
Comment (maxsokolovsky) [reply pkg/catalogv2/system/system.go:435]: 1. I don't believe so, since `isInstalled` receives the minVersion parameter. It's all about the value that the caller would pass in. In our case, `install` determines what `minVersion` means.

File: pkg/settings/setting.go (hunk 4 above)
Comment (maxsokolovsky) [reply pkg/settings/setting.go:171]: Agreed, though I would say we follow semantic versioning very loosely in our releases.
Should I have two environment variables (and thus two fields, etc.)? This could have problems if both are used and have different logic. But if we just keep the name but have a different meaning and logic, then we could do that.

Comment (maxsokolovsky) [reply pkg/settings/setting.go]: @KevinJoiner, I spoke with Caleb and Michael about this. We decided to keep the old setting but make it non-functional and mark as deprecated. It's to be removed in v2.8.
We'll add a `CATTLE_RANCHER_WEBHOOK_VERSION` environment variable. It should only be set in `package/Dockerfile`. Any code that needs to read it will do so using `os.Getenv`.

File: pkg/catalogv2/system/system.go (hunk 1 above)
Comment (MbolotSuse) [reply pkg/catalogv2/system/system.go:436]: I'm also a bit confused on the behavior here. In the following scenario:
- Min version is 2.0.1
- Installed is 2.0.2
- Exact version is "" (empty)

I would expect isInstalled to return `true`. However, this code seems to indicate that this would now return false, which may cause charts which use the min version to be re-installed. Would you mind clarifying this a bit more?

File: pkg/controllers/dashboard/systemcharts/controller.go [hunk 5]
```diff
@@ -101,7 +111,12 @@ func (h *handler) onRepo(key string, repo *catalog.ClusterRepo) (*catalog.Cluste
 		}
 		// webhook needs to be able to adopt the MutatingWebhookConfiguration which originally wasn't a part of the
 		// chart definition, but is now part of the chart definition
-		if err := h.manager.Ensure(chartDef.ReleaseNamespace, chartDef.ChartName, chartDef.MinVersionSetting.Get(), values, chartDef.ChartName == webhookChartName, installImageOverride); err != nil {
+		minVersion := chartDef.MinVersionSetting.Get()
+		exactVersion := chartDef.ExactVersionSetting.Get()
```
Comment (MbolotSuse) [inline pkg/controllers/dashboard/systemcharts/controller.go:115]: I think I get what you are going for here, but I think that this is a bit error prone for the future and could lead to nil pointers if someone has specified one of these values but not the other. I would recommend putting in a side function, something like the following:

```golang

// getChartVersion gets the min version and exact version (in that order) for a chart
def getChartVersions(chartDef chart.Definition) (string, string){
    minVersion := ""
    if chartDef.MinVersionSetting != nil{
        minVersion = chartDef.MinVersionSettting.Get()
    }
    if chartDef.ExactVersionSetting != nil{
        exactVersion = chartDef.ExactVersionSetting.Get()
    }
    return minVersion, exactVersion

}
```

File: pkg/controllers/dashboard/systemcharts/controller.go [hunk 6]
```diff
@@ -101,7 +111,12 @@ func (h *handler) onRepo(key string, repo *catalog.ClusterRepo) (*catalog.Cluste
...
+		if chartDef.ChartName == webhookChartName && minVersion != "" {
+			exactVersion = ""
```
Comment (MbolotSuse) [inline pkg/controllers/dashboard/systemcharts/controller.go:117]: Nit: Might be good to leave a comment here about how, for the webhook, we want the legacy parameter to take precedent.

File: pkg/catalogv2/system/system.go (hunk 3 above)
Comment (MbolotSuse) [reply pkg/catalogv2/system/system.go:233]: One thing to note here: this does effectively change the interface of this function. Before, you needed to know the previously request min version to remove from the desired charts. Now you just need to know the name/namespace. 

I think this is ok since the only function that was using remove seems to be the hosted stuff (eks/aks/gke), and it doesn't seem to specify a min version for install or remove (it is "" in both cases).

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Still looking at tests, but wanted to leave a partial review on the prod side.

File: pkg/controllers/dashboard/systemcharts/controller.go (hunk 5 above)
Comment (maxsokolovsky) [reply pkg/controllers/dashboard/systemcharts/controller.go:115]: I don't foresee nil pointers here, since `chart.Definition` stores a **value** of type `Setting`, not a pointer to it.
Also, I do have tests that set one value but not the other, and there are no nil pointers. Integration tests also have such a scenario.
https://github.com/rancher/rancher/blob/release/v2.7/pkg/controllers/dashboard/systemcharts/controller_test.go

Comment (MbolotSuse) [reply pkg/controllers/dashboard/systemcharts/controller.go:115]: Fair comment. As long as we are confident that "Setting{}.Get()" will return "", then I think we can resolve this.

File: tests/v2/integration/catalogv2/system_charts_version_test.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+		} else if event.Type == watch.Added {
+			return true, nil
+		}
+		return false, nil
+	})
+	w.Require().NoError(err)
+
+	// Allow the new release to fully deploy. Otherwise, the client won't find it among current releases.
+	var newRelease *release.Release
+	err = kwait.Poll(10*time.Second, time.Minute, func() (done bool, err error) {
```
Comment (MbolotSuse) [inline tests/v2/integration/catalogv2/system_charts_version_test.go:108]: It might be good to define the timeout of this in a constant at the top of the file, and include the timeout in the error message if things don't install in the right time. I'm thinking that the "if you change the version a new webhook installs within 'x'seconds" could be part of our functional contract.

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Couple of comments on the tests.

File: pkg/catalogv2/system/system.go (hunk 3 above)
Comment (KevinJoiner) [reply pkg/catalogv2/system/system.go:233]: This function is also not tested by integration tests.

File: pkg/catalogv2/system/system.go (hunk 1 above)
Comment (KevinJoiner) [reply pkg/catalogv2/system/system.go:436]: It would be helpful to add a unit test just for the `isInstalled` function. That way, we can make sure we cover all of the edge-cases, but don't have to refactor the entire file. If we need to, we can even break up the function and just test the logic we are changing.

File: pkg/controllers/dashboard/systemcharts/controller_test.go [hunk 8]
```diff
@@ -110,6 +114,47 @@ func Test_ChartInstallation(t *testing.T) {
...
+					gomock.AssignableToTypeOf(b),
+					"rancher-test.io/"+settings.ShellImage.Get(),
+				).Return(nil)
+
+				manager.EXPECT().Uninstall(operatorNamespace, "rancher-operator").Return(nil)
+				return manager
+			},
+			registryOverride: "rancher-test.io",
+		},
+		// TODO (maxsokolovsky) remove once the rancher-webhook-min-version setting is fully removed.
```
Comment (KevinJoiner) [inline pkg/controllers/dashboard/systemcharts/controller_test.go:151]: We can remove this TODO and create a new issue to remove it from the code. similar to https://github.com/rancher/rancher/issues/38535

File: pkg/rancher/rancher.go [hunk 9]
```diff
@@ -464,7 +464,7 @@ func setupRancherService(ctx context.Context, restConfig *rest.Config, httpsList
 // This should only be called when Rancher is run in a Docker container because the Kubernetes version and Rancher version
 // are bumped at the same time. In a Kubernetes cluster, usually the Rancher version is bumped when the cluster is upgraded.
 func bumpRancherWebhookIfNecessary(ctx context.Context, restConfig *rest.Config) error {
-	webhookVersionParts := strings.Split(os.Getenv("CATTLE_RANCHER_WEBHOOK_MIN_VERSION"), "+up")
+	webhookVersionParts := strings.Split(os.Getenv("CATTLE_RANCHER_WEBHOOK_VERSION"), "+up")
```
Comment (KevinJoiner) [inline pkg/rancher/rancher.go:467]: We don't need to update this value because this case shouldn't be hit on versions of Rancher > 2.7.3 because we do not support K8s 1.22+.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/rancher/rancher.go (hunk 9 above)
Comment (KevinJoiner) [inline pkg/rancher/rancher.go:467]:
```suggestion
	minVer := os.Getenv("CATTLE_RANCHER_WEBHOOK_MIN_VERSION")
	if minVer == "" {
		minVer = os.Getenv("CATTLE_RANCHER_WEBHOOK_VERSION")
	}
	webhookVersionParts := strings.Split(minVer, "+up")
```

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: tests/v2/integration/catalogv2/system_charts_version_test.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+	w.session = testSession
+
+	w.client, err = rancher.NewClient("", testSession)
+	require.NoError(w.T(), err)
+
+	w.catalogClient, err = w.client.GetClusterCatalogClient("local")
+	require.NoError(w.T(), err)
+
+	host := hostName(w.client.RestConfig.Host)
+	w.restClientGetter = &genericclioptions.ConfigFlags{
```
Comment (igomez06) [inline tests/v2/integration/catalogv2/system_charts_version_test.go:57]: Max instead of exporting the rest.Config we have extensions that do this.
  ```
kubeConfig, err := kubeconfig.GetKubeconfig(client, clusterID)
  require.NoError()
  restConfig, err := (*kubeConfig).ClientConfig()
  require.NoError()
 
 newClientGett := NewRestGetter(restConfig, clientConfig)
```
NewRestGetteer is in https://github.com/rancher/rancher/blob/release/v2.7/tests/framework/extensions/kubeconfig/restgetter.go

File: tests/v2/integration/catalogv2/system_charts_version_test.go [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+func TestSystemChartsVersionSuite(t *testing.T) {
+	suite.Run(t, new(SystemChartsVersionSuite))
+}
+
+func (w *SystemChartsVersionSuite) TestInstallWebhook() {
+	defer w.resetSettings()
+
+	const exactVersion = "2.0.3+up0.3.3"
```
Comment (igomez06) [inline tests/v2/integration/catalogv2/system_charts_version_test.go:86]: should we put consts at the top of the file.

File: tests/v2/integration/catalogv2/system_charts_version_test.go [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+	w.Require().NoError(w.uninstallApp("cattle-system", "rancher-webhook"))
+	w.Require().NoError(w.updateSetting("rancher-webhook-version", exactVersion))
+
+	watcher, err := w.catalogClient.Apps("cattle-system").Watch(context.TODO(), metav1.ListOptions{
```
[the remark this answers is not in the document]
Comment (igomez06) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:90]: yeah I think it makes sense to make the "waitForRollout" an extension do reduce duplication. Even look at our existing charts extensions. https://github.com/rancher/rancher/tree/release/v2.7/tests/framework/extensions/charts

File: tests/v2/integration/catalogv2/system_charts_version_test.go [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+	w.Require().NoError(err)
+
+	latest, err := w.catalogClient.GetLatestChartVersion("fleet")
+	w.Require().NoError(err)
+
+	// Ensure Rancher deployed the latest version when the minimum version is below the latest.
+	w.Assert().Equal(newRelease.Chart.Metadata.Version, latest)
+}
+
+func (w *SystemChartsVersionSuite) uninstallApp(namespace, chartName string) error {
```
Comment (igomez06) [inline tests/v2/integration/catalogv2/system_charts_version_test.go:204]: uninstallApp, getReleases, fetchReleases. Seem like good candidates for extensions, if we are to reuse this functionality for future chart tests.

**Review by igomez06 — COMMENTED** [review]
For the test side left comments, was curious if this test is fine to test against local, or against the downstream cluster.

File: tests/v2/integration/catalogv2/system_charts_version_test.go (hunk 11 above)
Comment (maxsokolovsky) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:86]: I don't think so, since they are scoped to individual tests.

File: tests/v2/integration/catalogv2/system_charts_version_test.go (hunk 13 above)
Comment (maxsokolovsky) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:204]: Agreed! I'd be glad to do it, but I might not have time to factor this out in this PR - this issue needs to be complete by EOD.

File: tests/v2/integration/catalogv2/system_charts_version_test.go (hunk 10 above)
Comment (maxsokolovsky) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:57]: Thanks for the link! For this to work, I might need some help implementing the `ToDiscoveryClient` method on the `RestGetter`. It gets called by Helm, and our getter just returns nil for that now.

File: pkg/catalogv2/system/system.go [hunk 14]
```diff
@@ -399,41 +406,47 @@ func (m *Manager) isInstalled(namespace, name, version, minVersion string, desir
...
 
-		ver, err := semver.NewVersion(release.Chart.Metadata.Version)
+		desired, err := semver.NewVersion(latestVersion)
 		if err != nil {
 			return false, "", nil, err
 		}
 
+		if minVersion == "" {
+			// Use exact version.
+			if !current.Equal(desired) {
```
Comment (MbolotSuse) [inline pkg/catalogv2/system/system.go:421]: This logic relies on having exactly one release. If there are other releases, it's possible that there could be an existing release which matches this version exactly but you didn't process that one first, so you returned false and requested a new install. 

Currently, no part of the code makes that assertion. If you can elaborate on your confidence level here (or why you think that this won't return more than one release on a list) that would be helpful.

File: pkg/controllers/dashboard/systemcharts/controller_test.go (hunk 8 above)
Comment (maxsokolovsky) [reply pkg/controllers/dashboard/systemcharts/controller_test.go:151]: Pasting the new issue here: https://github.com/rancher/rancher/issues/41561.

File: pkg/catalogv2/system/system.go [hunk 15]
```diff
@@ -399,41 +406,47 @@ func (m *Manager) isInstalled(namespace, name, version, minVersion string, desir
...
 			return false, "", nil, err
 		}
 
-		ver, err := semver.NewVersion(release.Chart.Metadata.Version)
+		desired, err := semver.NewVersion(latestVersion)
 		if err != nil {
 			return false, "", nil, err
 		}
 
+		if minVersion == "" {
```
Comment (MbolotSuse) [inline pkg/catalogv2/system/system.go:419]: Nit:

This part of the logic makes the exact version overrideable by the min version. While that is the behavior that we want for the webhook's min version, I think that there's an argument to be made that we don't want this to be enforced for everything - and that the webhook should choose, when requesting an install, to request either a min version or exact version install.

File: pkg/catalogv2/system/system_test.go
Comment (MbolotSuse) [inline pkg/catalogv2/system/system_test.go:1]: Thanks for adding tests for the new function. Can you also add tests for the `remove` function?

**Review by MbolotSuse — CHANGES_REQUESTED** [review]
Review of prod-side code, will look over tests as well.

File: pkg/catalogv2/system/system.go [hunk 16]
```diff
@@ -367,24 +372,26 @@ func (m *Manager) isInstalled(namespace, name, version, minVersion string, desir
...
-	if err != nil {
-		return false, "", nil, err
-	}
+	return isInstalled(releases, minVersion, version, desiredValue)
+}
 
-	for _, release := range releases {
-		if release.Info.Status != release2.StatusDeployed {
+func isInstalled(releases []*release.Release, minVersion, latestVersion string, desiredValues map[string]any) (bool, string, map[string]interface{}, error) {
+	if minVersion == "" && latestVersion == "" {
```
[the remark this answers is not in the document]
Comment (maxsokolovsky) [reply pkg/catalogv2/system/system.go:379]: Two things: explicit requirements for the inputs and performance.
Without the check, the behavior will be the same. The function will just throw an error later when attempting to parse the empty string as one denoting some semantic version. We don't need to let it go that far and should just check the basic case first.

File: pkg/catalogv2/system/system.go [hunk 17]
```diff
@@ -399,41 +406,47 @@ func (m *Manager) isInstalled(namespace, name, version, minVersion string, desir
...
 			}
-			if min.LessThan(ver) || min.Equal(ver) {
+			if min.LessThan(current) || min.Equal(current) {
 				// If the current deployed version is greater or equal than the min version but configuration has changed, return false and upgrade with the current version
 				if !bytes.Equal(patchedJSON, actualValueJSON) {
-					return false, release.Chart.Metadata.Version, desiredValue, nil
+					return false, r.Chart.Metadata.Version, desiredValues, nil
 				}
-				logrus.Debugf("Skipping installing/upgrading desired version %v for release %v, current version %v is greater or equal to minimal required version %v", desired.String(), name, ver.String(), minVersion)
+				logrus.Debugf("Skipping installing/upgrading desired version %v for release %v, current version %v is greater or equal to minimal required version %v", desired.String(), r.Chart.Name(), current.String(), minVersion)
```
Comment (MbolotSuse) [inline pkg/catalogv2/system/system.go:440]: Nit:

You are using `chart.Name()` here, but the prior version was using `name` which seems to be the release name (the log message also indicates that this is supposed to be the release name as well).

File: tests/v2/integration/catalogv2/system_charts_version_test.go (hunk 10 above)
Comment (igomez06) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:57]: Let me know!

File: pkg/catalogv2/system/system_test.go
Comment (maxsokolovsky) [reply pkg/catalogv2/system/system_test.go:1]: That method is an interface method, but I'd need to test the concrete implementation on the Manager value.

File: pkg/catalogv2/system/system.go (hunk 14 above)
Comment (KevinJoiner) [reply pkg/catalogv2/system/system.go:421]: 1. This is the same behavior as the previous version if minVersion equals empty string. ( I wrote some additional tests to verify this.)
2. This logic relies on only one version of a release being deployed at a time which is enforced by Helm. The loop is to handle skipping over releases that are not deployed.

File: pkg/catalogv2/system/system_test.go
Comment (MbolotSuse) [reply pkg/catalogv2/system/system_test.go:1]: Right, that's ok.

File: pkg/catalogv2/system/system_test.go [hunk 18]
```diff
@@ -38,3 +41,295 @@ func TestGetIntervalOrDefault(t *testing.T) {
...
+			name:          "min and latest are both unset",
+			latestVersion: "",
+			minVersion:    "",
+			desiredValues: standardValues,
+
+			expectedInstalled: false,
+			expectedVersion:   "",
+			expectedValues:    nil,
+			expectedErr:       true,
+		},
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system_test.go:219]: Can you add these test cases
``` golang
		{
			name:          "no min version latest version is greater",
			latestVersion: "2.0.0",
			minVersion:    "",
			desiredValues: standardValues,

			expectedInstalled: false,
			expectedVersion:   "2.0.0",
			expectedValues:    standardValues,
			expectedErr:       false,
		},
		{
			name:          "no min version latest version equals current",
			latestVersion: "1.0.0",
			minVersion:    "",
			desiredValues: standardValues,

			expectedInstalled: true,
			expectedVersion:   "",
			expectedValues:    nil,
			expectedErr:       false,
		},
		{
			name:          "no min version latest version equals current values differ",
			latestVersion: "1.0.0",
			minVersion:    "",
			desiredValues: newValues,

			expectedInstalled: false,
			expectedVersion:   "1.0.0",
			expectedValues:    newValues,
			expectedErr:       false,
		},
```

File: pkg/catalogv2/system/system_test.go [hunk 19]
```diff
@@ -38,3 +41,295 @@ func TestGetIntervalOrDefault(t *testing.T) {
...
+			Config: standardValues,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			// Note that the minVersion argument must be an empty string.
+			installed, version, values, err := isInstalled(releases, "", test.exactVersion, test.desiredValues)
```
Comment (KevinJoiner) [inline pkg/catalogv2/system/system_test.go:328]: If we are not explicitly enforcing this behavior, we may need to test the interaction between setting both exactVersion and minVersion.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]

File: pkg/catalogv2/system/system.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (MbolotSuse) [reply pkg/catalogv2/system/system.go:379]: > explicit requirements for the inputs and performance.

IMO, the `latestVersion` parameter isn't controlled by the user, so having an explicit requirement for this value doesn't make too much sense. Since this value is derived from the chart itself, there's various reasons (some of which are in the calling function's control control, some of which aren't) which could be causing an issue here. 

If we want a requirement for this, I think it would best go in Ensure so that the function which calls ensure get's an error that it supplied invalid input.

File: pkg/catalogv2/system/system_test.go (hunk 18 above)
Comment (maxsokolovsky) [reply pkg/catalogv2/system/system_test.go:219]: I already have these cases in a separate test: `TestIsInstalledExactVersion`.

**Comment by maxsokolovsky** [comment]
Spoke with @MbolotSuse and @KevinJoiner. We decided to fix one regression (the new code no longer uses a merged version of old and new values) and aim to merge this PR. Later, we will have another issue whose goal is to have a more substantial refactoring.

**Review by KevinJoiner — DISMISSED** [review]

**Review by MbolotSuse — APPROVED** [review]

**Review by MbolotSuse — APPROVED** [review]

**Review by KevinJoiner — APPROVED** [review]
🏆

File: tests/v2/integration/catalogv2/system_charts_version_test.go (hunk 10 above)
Comment (igomez06) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:57]: Hey Max I know this needs to be merged, but to avoid making a pretty fundamental change to the framework could you get the restGetter to work by get the restConfig like this instead?
```
kubeConfig, err := kubeconfig.GetKubeconfig(client, "local")
require.NoError()
restConfig, err := (*kubeConfig).ClientConfig()
require.NoError()

```

Comment (maxsokolovsky) [reply tests/v2/integration/catalogv2/system_charts_version_test.go:57]: Yup, it's done, will open a follow-up PR soon.
